src/coreclr/jit/gentree.cpp

```diff
@@ -24133,6 +24165,390 @@ GenTree* Compiler::gtNewSimdSumNode(var_types type, GenTree* op1, CorInfoType si
 #endif // !TARGET_XARCH && !TARGET_ARM64
 }
 
+#if defined(TARGET_XARCH)
+GenTree* Compiler::gtNewSimdTernaryLogicNode(var_types   type,
+                                             GenTree*    op1,
+                                             GenTree*    op2,
+                                             GenTree*    op3,
+                                             GenTree*    op4,
+                                             CorInfoType simdBaseJitType,
+                                             unsigned    simdSize)
+{
+    assert(IsBaselineVector512IsaSupportedDebugOnly());
+
+    assert(varTypeIsSIMD(type));
+    assert(getSIMDTypeForSize(simdSize) == type);
+
+    assert(op1 != nullptr);
+    assert(op1->TypeIs(type));
+
+    assert(op2 != nullptr);
+    assert(op2->TypeIs(type));
+
+    assert(op3 != nullptr);
+    assert(op3->TypeIs(type));
+
+    assert(op4 != nullptr);
+    assert(genActualType(op4) == TYP_INT);
+
+    var_types simdBaseType = JitType2PreciseVarType(simdBaseJitType);
+    assert(varTypeIsArithmetic(simdBaseType));
+
+    NamedIntrinsic intrinsic = NI_Illegal;
+
+    if (simdSize == 64)
+    {
+        intrinsic = NI_AVX512F_TernaryLogic;
+    }
+    else
+    {
+        assert((simdSize == 16) || (simdSize == 32));
+        intrinsic = NI_AVX512F_VL_TernaryLogic;
+    }
+
+    if (op4->IsIntegralConst())
+    {
+        // We have a constant control byte, so we can potentially optimize
+
+        uint8_t                 control  = static_cast<uint8_t>(op4->AsIntCon()->gtIconVal);
+        const TernaryLogicInfo& info     = TernaryLogicInfo::lookup(control);
+        TernaryLogicUseFlags    useFlags = info.GetAllUseFlags();
+
+        if (useFlags != TernaryLogicUseFlags::ABC)
+        {
+            // We are not using all 3 inputs, so we can potentially optimize
+
+            assert(info.oper2 != TernaryLogicOperKind::Select);
+            assert(info.oper2 != TernaryLogicOperKind::True);
+            assert(info.oper2 != TernaryLogicOperKind::False);
+            assert(info.oper2 != TernaryLogicOperKind::Cond);
+            assert(info.oper2 != TernaryLogicOperKind::Major);
+            assert(info.oper2 != TernaryLogicOperKind::Minor);
+            assert(info.oper3 == TernaryLogicOperKind::None);
+            assert(info.oper3Use == TernaryLogicUseFlags::None);
+
+            GenTree* value1;
+            GenTree* value2;
+
+            switch (useFlags)
+            {
+                case TernaryLogicUseFlags::A:
+                {
+                    // One operand. We swap it with C so it can be contained where possible
+                    value1 = op1;
+                    value2 = nullptr;
+
+                    std::swap(op1, op3);
```

Comment:
This is also problematic, we cannot just swap these around if they can interfere in any way (with each other or with `op2`).

Reply:
How would you recommend this be handled? In cases like `AndNot` we require they have been properly spilled before calling `gtNewSimdBinOpNode(GT_AND_NOT, ...)`

In this case, we can't easily know ahead of time if reordering will be needed without duplicating some of the checks/handling

Reply:
I've duplicated some of the logic and added asserts that there are no side effects before the swap for the time being.

Reply:
I'd suggest leaving the folding logic out of this PR and introducing it in the future if/when we see actual cases where it would be beneficial. It's 500+ lines of complex logic that seems of questionable benefit to me and with high risk of being incorrect. I don't see a simple way to make it handle side effects properly in the setting of a constructor that's supposed to be usable from anywhere in the JIT -- requiring that the operands do not have any side effects and are able to be freely reordered is very unusual for a constructor.

Reply:
This logic is important for correctly optimizing the cases I already called out, such as `NOT`, `NAND`, `NOR`, and `XNOR`.

It's not something we can just leave out.

Reply:
Updated the logic based on the discussion we had in Discord.

The complex logic has been moved into `import` instead since the JIT will never be producing such nodes itself. A decently large explanatory comment was added to the import logic to help explain what we're doing and why.